cluster_describer.go

```diff
@@ -0,0 +1,65 @@
+package kafka
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/kafka"
+	"github.com/aws/aws-sdk-go-v2/service/kafka/types"
+)
+
+const (
+	clusterConfigTimeout      = 5 * time.Second
+	clusterDescriptionTimeout = 10 * time.Second
+)
+
+type clusterDescriber interface {
+	DescribeClusterV2(ctx context.Context, input *kafka.DescribeClusterV2Input, optFns ...func(*kafka.Options)) (*kafka.DescribeClusterV2Output, error)
+}
+
+func newClusterDescriber() (clusterDescriber, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), clusterConfigTimeout)
+	defer cancel()
+
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return kafka.NewFromConfig(cfg), nil
+}
+
+// Verifies whether the Kafka cluster is available or not.
+// False positive healthcheck errors are being ignored during maintenance windows.
+func checkClusterAvailability(healthErr error, describer clusterDescriber, arn *string) error {
+	state, stateErr := retrieveClusterState(describer, arn)
+	if stateErr != nil {
+		return fmt.Errorf("cluster status is unknown: %w", stateErr)
+	}
+
+	if isMaintenanceState(state) {
+		return nil
+	}
+
+	return healthErr
+}
+
+func retrieveClusterState(describer clusterDescriber, arn *string) (types.ClusterState, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), clusterDescriptionTimeout)
+	defer cancel()
+
+	cluster, err := describer.DescribeClusterV2(ctx, &kafka.DescribeClusterV2Input{
```

Comment:
**nitpick (non-blocking):** A very theoretical scenario would be for `newClusterDescriber()` to return an error when created in the `NewConsumer` or `NewProducer`. When using `checkClusterAvailability` you only check for available `config.ClusterArn`. Again theoretically `describer` could be nil here.

Reply:
The `NewConsumer` and `NewProducer` functions will only attempt to create a describer if `ClusterArn` is provided. If that fails, they'd return an error which would cause the termination of the whole application. By doing this the describer is never going to be nil.

Then I opted for checking `ClusterArn` just because it's simpler than nil checking an interface. Do you think it's better to check the describer instead?

Reply:
Okay, I understand the idea now, my bad. I think it's fine that way it's implemented having in mind that the whole logic is private.
If `retrieveClusterState` was public and we wanted to be super foolproof we can check if the describer is not null before `describer.DescribeClusterV2` call.